File: examples/php/order.php
<?php
/*
 * Copyright 2026 Chengxi Luo
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * `OrderRecord` mirrors examples/cases/order.yaml — a placed order.
 *
 * `LineItem` mirrors the reusable line_item.yaml it imports, which itself nests
 * money — so any type using it exercises struct-inside-struct. `Address` and
 * `Money` come from customer.php, as they do in the Go worker.
 *
 * Between them the fields cover the four composite types nothing else in the
 * suite exercises end to end: an `enum`, a `list<struct>`, a
 * `map<string,struct>` and an `optional<struct>`. `$billingAddress` is the
 * suite's only optional<struct>, and unlike the list and map beside it, it
 * needs no `elem:` — its own property type says which class it is.
 *
 * An enum needs nothing from the binding: it travels as its variant *name*, so
 * the property is a plain string. The u8 ordinal in the layout is this worker's
 * own choice, which is why STATUSES has to match the case file's declaration
 * order.
 *
 * PHP's int is 64-bit *signed*, so the two u64 ids travel as decimal strings
 * through ext-gmp, as the other models here do.
 *
 * Go is the --ref language and owns the layout; see examples/go/wire.go.
 */

declare(strict_types=1);

use Serify\Attributes\SerifyField;
use Serify\Attributes\SerifyModel;

/** Declaration order of the `status` enum in examples/cases/order.yaml. */
const ORDER_STATUSES = ['pending', 'paid', 'shipped', 'delivered', 'cancelled'];

#[SerifyModel]
class LineItem
{
    #[SerifyField] public string $sku = '';
    #[SerifyField] public string $productName = '';
    #[SerifyField] public int $quantity = 0;
    #[SerifyField] public Money $unitPrice;
    #[SerifyField] public int $discountPct = 0;
    #[SerifyField] public bool $giftWrap = false;

    public function __construct()
    {
        $this->unitPrice = new Money();
    }

    public function pack(): string
    {
        return lenPrefixed($this->sku)
            . lenPrefixed($this->productName)
            . pack('v', $this->quantity)
            . $this->unitPrice->pack()
            . pack('CC', $this->discountPct, $this->giftWrap ? 1 : 0);
    }

    public static function unpack(string $data, int &$off): self
    {
        $it = new self();
        $it->sku = readLenPrefixed($data, $off);
        $it->productName = readLenPrefixed($data, $off);
        $it->quantity = unpack('v', $data, $off)[1];
        $off += 2;
        $it->unitPrice = Money::unpack($data, $off);
        $it->discountPct = unpack('C', $data, $off)[1];
        $it->giftWrap = unpack('C', $data, $off + 1)[1] !== 0;
        $off += 2;
        return $it;
    }
}

#[SerifyModel]
class OrderRecord
{
    #[SerifyField] public string $orderId = '0';
    #[SerifyField] public string $customerId = '0';
    #[SerifyField] public string $createdAt = '0';
    #[SerifyField] public string $status = '';
    #[SerifyField(elem: LineItem::class)] public array $items = [];
    #[SerifyField] public Money $subtotal;
    #[SerifyField(elem: Money::class)] public array $adjustments = [];
    #[SerifyField] public Money $total;
    #[SerifyField] public Address $shippingAddress;
    #[SerifyField] public ?Address $billingAddress = null;
    #[SerifyField] public array $couponCodes = [];
    #[SerifyField] public ?string $trackingNumber = null;

    public function __construct()
    {
        $this->subtotal = new Money();
        $this->total = new Money();
        $this->shippingAddress = new Address();
    }

    public function marshal(): string
    {
        $out = encodeInt($this->orderId, 8)
            . encodeInt($this->customerId, 8)
            . encodeInt($this->createdAt, 8);

        // enum: a u8 ordinal, the variant's position in the case file.
        $ord = array_search($this->status, ORDER_STATUSES, true);
        if ($ord === false) {
            throw new RuntimeException("unknown order status \"{$this->status}\"");
        }
        $out .= pack('C', $ord);

        $out .= pack('V', count($this->items));
        foreach ($this->items as $it) {
            $out .= $it->pack();
        }

        $out .= $this->subtotal->pack();

        // Entry order is the array's own — deliberately not sorted. A map is
        // unordered, so order declares `oracle: semantic` and the decoded value
        // is what gets compared. See docs/protocol.md.
        $out .= pack('V', count($this->adjustments));
        foreach ($this->adjustments as $k => $m) {
            $out .= lenPrefixed((string) $k) . $m->pack();
        }

        $out .= $this->total->pack();
        $out .= $this->shippingAddress->pack();

        // optional<struct>: a presence flag, then the struct's fields inline.
        if ($this->billingAddress === null) {
            $out .= pack('C', 0);
        } else {
            $out .= pack('C', 1) . $this->billingAddress->pack();
        }

        $out .= pack('V', count($this->couponCodes));
        foreach ($this->couponCodes as $c) {
            $out .= lenPrefixed($c);
        }

        if ($this->trackingNumber === null) {
            $out .= pack('C', 0);
        } else {
            $out .= pack('C', 1) . lenPrefixed($this->trackingNumber);
        }

        return $out;
    }

    public static function unmarshal(string $data): self
    {
        $o = new self();

        $o->orderId = decodeUnsigned(substr($data, 0, 8));
        $o->customerId = decodeUnsigned(substr($data, 8, 8));
        $o->createdAt = decodeSigned(substr($data, 16, 8));

        $ord = unpack('C', $data, 24)[1];
        if (!isset(ORDER_STATUSES[$ord])) {
            throw new RuntimeException("status ordinal $ord is out of range");
        }
        $o->status = ORDER_STATUSES[$ord];
        $off = 25;

        $o->items = [];
        for ($n = self::readCount($data, $off); $n > 0; $n--) {
            $o->items[] = LineItem::unpack($data, $off);
        }

        $o->subtotal = Money::unpack($data, $off);

        $o->adjustments = [];
        for ($n = self::readCount($data, $off); $n > 0; $n--) {
            $k = readLenPrefixed($data, $off);
            $o->adjustments[$k] = Money::unpack($data, $off);
        }

        $o->total = Money::unpack($data, $off);
        $o->shippingAddress = Address::unpack($data, $off);

        if (unpack('C', $data, $off)[1] === 0) {
            $o->billingAddress = null;
            $off += 1;
        } else {
            $off += 1;
            $o->billingAddress = Address::unpack($data, $off);
        }

        $o->couponCodes = [];
        for ($n = self::readCount($data, $off); $n > 0; $n--) {
            $o->couponCodes[] = readLenPrefixed($data, $off);
        }

        if (unpack('C', $data, $off)[1] === 0) {
            $o->trackingNumber = null;
        } else {
            $off += 1;
            $o->trackingNumber = readLenPrefixed($data, $off);
        }

        return $o;
    }

    private static function readCount(string $data, int &$off): int
    {
        $n = unpack('V', $data, $off)[1];
        $off += 4;
        return $n;
    }
}

File: examples/elixir/lib/order.ex
# Copyright 2026 Chengxi Luo
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

defmodule LineItem do
  @moduledoc """
  Mirrors the reusable examples/cases/line_item.yaml, imported by order.

  Its own `unit_price` is a money, which makes any type using it the suite's
  only struct-inside-struct.
  """
  use WorkerLib.Serify.Model

  alias Customer.Wire

  defstruct sku: "",
            product_name: "",
            quantity: 0,
            unit_price: %Money{},
            discount_pct: 0,
            gift_wrap: false

  serify_field(:sku, :string)
  serify_field(:product_name, :string)
  serify_field(:quantity, :u16)
  serify_field(:unit_price, :struct, module: Money)
  serify_field(:discount_pct, :u8)
  serify_field(:gift_wrap, :bool)

  def pack(%__MODULE__{} = it) do
    Wire.str(it.sku) <>
      Wire.str(it.product_name) <>
      <<it.quantity::little-16>> <>
      Money.pack(it.unit_price) <>
      <<it.discount_pct::8, if(it.gift_wrap, do: 1, else: 0)::8>>
  end

  def unpack(data) do
    {sku, data} = Wire.take_str(data)
    {product_name, <<quantity::little-16, rest::binary>>} = Wire.take_str(data)
    {unit_price, <<discount_pct::8, gift_wrap::8, rest::binary>>} = Money.unpack(rest)

    {%__MODULE__{
       sku: sku,
       product_name: product_name,
       quantity: quantity,
       unit_price: unit_price,
       discount_pct: discount_pct,
       gift_wrap: gift_wrap != 0
     }, rest}
  end
end

defmodule OrderRecord do
  @moduledoc """
  Mirrors examples/cases/order.yaml — a placed order.

  Between them the fields cover the four composite types nothing else in the
  suite exercises end to end: an `enum`, a `list<struct>`, a
  `map<string,struct>` and an `optional<struct>`. `billing_address` is the
  suite's only optional<struct>, so this is the first exercise of the binding's
  `{:optional, :struct}` clause.

  An enum needs nothing from the binding: it travels as its variant *name*, so
  the field is a plain string. The u8 ordinal in the layout is this worker's own
  choice, which is why @statuses has to match the case file's declaration order.

  Go is the --ref language and owns the layout; see examples/go/wire.go.
  """
  use WorkerLib.Serify.Model

  alias Customer.Wire

  # Declaration order of the `status` enum in examples/cases/order.yaml.
  @statuses ["pending", "paid", "shipped", "delivered", "cancelled"]

  defstruct order_id: 0,
            customer_id: 0,
            created_at: 0,
            status: "",
            items: [],
            subtotal: %Money{},
            adjustments: %{},
            total: %Money{},
            shipping_address: %Address{},
            billing_address: nil,
            coupon_codes: [],
            tracking_number: nil

  serify_field(:order_id, :u64)
  serify_field(:customer_id, :u64)
  serify_field(:created_at, :i64)
  serify_field(:status, :string)
  serify_field(:items, {:list, :struct}, module: LineItem)
  serify_field(:subtotal, :struct, module: Money)
  serify_field(:adjustments, :map, of: {:struct, Money})
  serify_field(:total, :struct, module: Money)
  serify_field(:shipping_address, :struct, module: Address)
  serify_field(:billing_address, {:optional, :struct}, module: Address)
  serify_field(:coupon_codes, {:list, :string})
  serify_field(:tracking_number, {:optional, :string})

  def marshal(%__MODULE__{} = o) do
    # enum: a u8 ordinal, the variant's position in the case file.
    ord =
      case Enum.find_index(@statuses, &(&1 == o.status)) do
        nil -> raise ArgumentError, "unknown order status #{inspect(o.status)}"
        i -> i
      end

    # Entry order is the map's own — deliberately not sorted. A map is
    # unordered, so order declares `oracle: semantic` and the decoded value is
    # what gets compared. See docs/protocol.md.
    adjustments =
      Enum.map_join(o.adjustments, fn {k, m} -> Wire.str(k) <> Money.pack(m) end)

    # optional<struct>: a presence flag, then the struct's fields inline.
    billing =
      case o.billing_address do
        nil -> <<0>>
        a -> <<1>> <> Address.pack(a)
      end

    tracking =
      case o.tracking_number do
        nil -> <<0>>
        s -> <<1>> <> Wire.str(s)
      end

    <<o.order_id::little-64, o.customer_id::little-64, o.created_at::signed-little-64, ord::8,
      length(o.items)::little-32>> <>
      Enum.map_join(o.items, &LineItem.pack/1) <>
      Money.pack(o.subtotal) <>
      <<map_size(o.adjustments)::little-32>> <>
      adjustments <>
      Money.pack(o.total) <>
      Address.pack(o.shipping_address) <>
      billing <>
      <<length(o.coupon_codes)::little-32>> <>
      Enum.map_join(o.coupon_codes, &Wire.str/1) <> tracking
  end

  def unmarshal(data) do
    <<order_id::little-64, customer_id::little-64, created_at::signed-little-64, ord::8,
      n_items::little-32, rest::binary>> = data

    status = Enum.at(@statuses, ord) || raise ArgumentError, "status ordinal #{ord} out of range"

    {items, rest} = Wire.take_n(rest, n_items, &LineItem.unpack/1)
    {subtotal, <<n_adj::little-32, rest::binary>>} = Money.unpack(rest)

    {adj_pairs, rest} =
      Wire.take_n(rest, n_adj, fn d ->
        {k, d} = Wire.take_str(d)
        {m, d} = Money.unpack(d)
        {{k, m}, d}
      end)

    {total, rest} = Money.unpack(rest)
    {shipping_address, <<has_billing::8, rest::binary>>} = Address.unpack(rest)

    {billing_address, rest} =
      if has_billing == 0, do: {nil, rest}, else: Address.unpack(rest)

    <<n_coupons::little-32, rest::binary>> = rest

    {coupon_codes, <<has_tracking::8, rest::binary>>} =
      Wire.take_n(rest, n_coupons, &Wire.take_str/1)

    {tracking_number, _rest} =
      if has_tracking == 0, do: {nil, rest}, else: Wire.take_str(rest)

    %__MODULE__{
      order_id: order_id,
      customer_id: customer_id,
      created_at: created_at,
      status: status,
      items: items,
      subtotal: subtotal,
      adjustments: Map.new(adj_pairs),
      total: total,
      shipping_address: shipping_address,
      billing_address: billing_address,
      coupon_codes: coupon_codes,
      tracking_number: tracking_number
    }
  end
end
